src/Avalonia.Interactivity/EventRoute.cs

```diff
@@ -0,0 +1,200 @@
+        {
+            target = target ?? throw new ArgumentNullException(nameof(target));
+
+            _route ??= new PooledList<RouteItem>(16);
+            _route.Add(new RouteItem(target, null, null, 0, false));
+        }
+
+        /// <summary>
+        /// Raises an event along the route.
+        /// </summary>
+        /// <param name="source">The event source.</param>
+        /// <param name="e">The event args.</param>
+        public void RaiseEvent(IInteractive source, RoutedEventArgs e)
+        {
+            source = source ?? throw new ArgumentNullException(nameof(source));
+            e = e ?? throw new ArgumentNullException(nameof(e));
+
+            e.Source = source;
+
+            if (_event.RoutingStrategies == RoutingStrategies.Direct)
+            {
+                e.Route = RoutingStrategies.Direct;
+                RaiseEventImpl(e);
+                _event.InvokeRouteFinished(e);
+            }
+            else
+            {
+                if (_event.RoutingStrategies.HasFlagCustom(RoutingStrategies.Tunnel))
+                {
+                    e.Route = RoutingStrategies.Tunnel;
+                    RaiseEventImpl(e);
+                    _event.InvokeRouteFinished(e);
+                }
+
+                if (_event.RoutingStrategies.HasFlagCustom(RoutingStrategies.Bubble))
+                {
+                    e.Route = RoutingStrategies.Bubble;
+                    RaiseEventImpl(e);
+                    _event.InvokeRouteFinished(e);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Disposes of the event route.
+        /// </summary>
+        public void Dispose()
+        {
+            _route?.Dispose();
+            _route = null;
+        }
+
+        private void RaiseEventImpl(RoutedEventArgs e)
+        {
+            if (_route is null)
+            {
+                return;
+            }
+
+            if (e.Source is null)
+            {
+                throw new ArgumentException("Event source may not be null", nameof(e));
+            }
+
+            IInteractive? lastTarget = null;
+            var start = 0;
+            var end = _route.Count;
+            var step = 1;
+
+            if (e.Route == RoutingStrategies.Tunnel)
+            {
+                start = end - 1;
+                step = end = -1;
+            }
+
+            for (var i = start; i != end; i += step)
+            {
+                var entry = _route[i];
+
+                // If we've got to a new control then call any RoutedEvent.Raised listeners.
+                if (entry.Target != lastTarget)
+                {
+                    if (!e.Handled)
+                    {
+                        _event.InvokeRaised(entry.Target, e);
+                    }
+
+                    // If this is a direct event and we've already raised events then we're finished.
+                    if (e.Route == RoutingStrategies.Direct && lastTarget is object)
```

Comment:
Since I've seen that in many of your recent PRs: what is wrong with the good old `!= null` ?

Reply:
It came from [this tweet](https://twitter.com/jaredpar/status/1115019017297596416) and the fact that C#'s equality semantics for structs are IMO broken.

When I was working on #3469 I changed `IndexPath` from a `class` to a `struct`. Because it override `==` that made every `== null` comparison break because when a struct overrides `==` it automatically stops `== null` being a compile error and just silently makes it always return `false`.

https://stackoverflow.com/questions/9186632/equality-operator-overloading-in-struct-and-classes